sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessorSpec.scala

```diff
@@ -0,0 +1,515 @@
+        // TEST_EVIDENCE: Integrity: well formed exercise-by-interface API command is accepted
+        validApiExeByInterface,
+        validApiExeByInterface.copy(typeRef = IName),
+        validApiExeByInterface.copy(argument = normalizedRecordNone),
+        validApiExeByInterface.copy(argument = nonNormalizedRecordNone),
+        validApiExeByInterface.copy(argument = nonNormalizedRecordUpgrade),
+        // TEST_EVIDENCE: Integrity: well formed exercise-by-key API command is accepted
+        validApiExeByKey,
+        validApiExeByKey.copy(templateRef = TName),
+        validApiExeByKey.copy(argument = nonNormalizedRecordNone),
+        validApiExeByKey.copy(argument = nonNormalizedRecordNone),
+        validApiExeByKey.copy(argument = nonNormalizedRecordNone),
+        validApiExeByKey.copy(contractKey = normalizedRecordNone),
+        validApiExeByKey.copy(contractKey = nonNormalizedRecordNone),
+        validApiExeByKey.copy(contractKey = nonNormalizedRecordUpgrade),
+        // TEST_EVIDENCE: Integrity: well formed exercise-by-key API command is accepted
+        validApiCreateAndExe,
+        validApiCreateAndExe.copy(templateRef = TName),
+        validApiCreateAndExe.copy(createArgument = normalizedRecordNone),
+        validApiCreateAndExe.copy(createArgument = nonNormalizedRecordNone),
+        validApiCreateAndExe.copy(createArgument = nonNormalizedRecordUpgrade),
+        validApiCreateAndExe.copy(choiceArgument = normalizedRecordNone),
+        validApiCreateAndExe.copy(choiceArgument = nonNormalizedRecordNone),
+        validApiCreateAndExe.copy(choiceArgument = nonNormalizedRecordUpgrade),
+      )
+
+      forEvery(testCases) { command =>
+        Try(
+          defaultPreprocessor.unsafePreprocessApiCommand(Map(pkg.pkgName -> pkgId), command)
+        ) shouldBe a[Success[_]]
+      }
+    }
+
+    "reject ill-formed API Commands" in {
+
+      val testCases = Table[ApiCommand, ResultOfATypeInvocation[_]](
+        ("command", "error"),
+        // TEST_EVIDENCE: Integrity: ill-formed create API command is rejected
+        validApiCreate.copy(templateRef = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validApiCreate.copy(templateRef = "Mod:I") ->
+          a[Error.Preprocessing.Lookup],
+        validApiCreate.copy(argument = illTypedRecord) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validApiCreate.copy(argument = invalidUpgradeRecord) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validApiCreate.copy(argument = withInvalidCidRecord) ->
+          a[Error.Preprocessing.IllegalContractId],
+        // TEST_EVIDENCE: Integrity: ill-formed exercise API command is rejected
+        validApiExe.copy(typeRef = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validApiExe.copy(typeRef = "Mod:S") ->
+          a[Error.Preprocessing.Lookup],
+        validApiExe.copy(choiceId = "Undefined") ->
```

Comment:
We could also have a case that exercises an existing choice that doesn't belong to the template.

Reply:
This is the case line 225